luigi/contrib/dropbox.py

```diff
@@ -0,0 +1,326 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2019 Jose-Ignacio Riaño Chico
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not
+# use this file except in compliance with the License. You may obtain a copy of
+# the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+
+from __future__ import absolute_import
+
+import logging
+import re
+import ntpath
+import os
+import random
+import tempfile
+import time
+from contextlib import contextmanager
+from functools import wraps
+
+import luigi.format
+from luigi.target import FileSystem, FileSystemTarget, AtomicLocalFile
+
+logger = logging.getLogger('luigi-interface')
+
+try:
+    import dropbox.dropbox
+    import dropbox.exceptions
+    import dropbox.files
+except ImportError:
+    logger.warning("Loading Dropbox module without the python package dropbox (https://pypi.org/project/dropbox/). "
+                   "Will crash at runtime if Dropbox functionality is used.")
+
+
+def accept_trailing_slash_in_existing_dirpaths(func):
+    @wraps(func)
+    def wrapped(self, path, *args, **kwargs):
+        if path != '/' and path.endswith('/') and self._exists_and_is_dir(re.sub("/+$", '', path)):
+            orig_path = path
+            path = re.sub("/+$", '', path)
```

Comment:
why use regex when you could just `path = path[:-1]`? much simpler

Reply:
looking at your tests, you're removing all trailing slashes, even if more than 1. if someone provides input of multiple trailing slashes, i'd expect that to error due to illogical input. whereas the single trailing slash is standard elsewhere (thus why we'd like to allow that format)

Reply:
I have changed the code to allow for only zero or one trailing slashes.

Initially, I thought about allowing multiple trailing slashes, but I have just noticed that Dropbox has a 'non-standardized' behavior when there are multiple slashes (e.g. they work for `files_list_folder()`, but not for `files_get_metadata()`. See image).

Therefore, now I think that you are right: allowing for only one trailing slash is the best option. 

![image](https://user-images.githubusercontent.com/4805351/58177705-27c23d80-7ca5-11e9-8f99-eb24cca150e8.png)
.

However, using paths with trailing slashes is still not a good idea, because it requires an additional API call which lowers the performance.

I have modified the documentation to reflect this (previously, it was `path must not end with '/'`; now it is `path should not end with '/'`) and also to make it explicit that adjacent slashes are not supported:

```
The path must start with '/' and should not end with '/' (even if it is a directory).
The path must not contain adjacent slashes ('/files//img.jpg' is an invalid path)
```
The must/should usage is [RFC-like style](https://www.ietf.org/rfc/rfc2119.txt), but if you prefer any other wording, I can change it.